### PR TITLE
"Security Implant" rework, prisoner management console updates

### PR DESCRIPTION
**Pull request by Rhials**
## About The Pull Request

For the vernacular purposes of the following PR body -- "Security Implant" refers to the existing subset of implants given, by security, to captured prisoners and such as a punitive, controlling measure. This includes the chemical, tracking, and maybe exile implants.

This revamps the functionality of how "security" implants are displayed on huds, prisoner management console implant controls/readouts, and their instrumentality. It was also, ultimately, an attempt at nerfing the tracking implant that spiralled far out of control.

Rather than only displaying chemical on the right and tracking on the left, all implants with the "security implant" flag will be trackable on SecHuds. A maximum of two can be implanted at once. This is both due to technical limitations, but also conveniently provides security a limit to consider when choosing implants.

Implants now also occupy their HUD slot based on the order they were implanted in, rather than always occupying the same spot. Neat!

image

From two (three if you count the exile implant), there are now five security implants. _The tracker implant has been split into two of these implants._

<details>
<summary>Summary of the implants, functions, changes:</summary>
<br>

- **Tracker (Red)** -- No longer grants teleporter beacon. Tracking radius has been increased from 20 to 35 tiles. The Prisoner Management Console will now list the area the prisoner is occupying as well. Disables after the implantee is dead for 10 minutes.
- **Chemical (Blue)** -- No mechanical changes. The implant pad readout has been modified slightly.
- **Exile (Green)** -- In addition to past functionality, station shuttle controls (public, mining, etc.) will be unresponsive for the implantee. Flimsy, but more effective than a stern warning not to come back from lavaland.
- **Beacon (Yellow)** -- Implantee becomes a teleporter beacon. The prisoner console will report if their currently occupied area is hazardous or not, so half of the security team doesn't blindly teleport into space or lava. Disables after the implantee is dead for 10 minutes. Available from Cargo.
- **Teleport Blocker (Deep Blue, not shown)** -- Prevents the implantee from being teleported. Ever wanted to keep a wizard or cultist in a cell? This is where you can start. Available from Cargo, expensive and scarce.

Each of the implants has some application that would benefit security if used on a captured criminal. Their usefulness may overlap in some places, but the overall range of control these implants give security is broadened.

</details>

The implant control console has also been given a small facelift. Certain implants provide more useful readouts that can help officers locate, control, or capture an implantee, rewarding cooperation between officers.

It has also been totally converted into TGUI by @MrMelbert. Kickass!

Also, You can now remotely destroy implants, either to relieve criminals from their punishment or to make room for a different implant. Wardens should keep hold of their ID and remember to log out, since a motivated convict could use it to shed their implants!

tgui

Everything made in this PR _should_ be scaleable enough to allow for new security implant types to be implemented with relative ease. The teleport-blocker implant was a last minute attempt to prove it to myself. I had a few more ideas for implants in my head, but figured this PR was already getting big and ugly enough. That is all for another day.

I truly apologize if there's anything I've missed in here. I did a lot of this over a long period of time and kind of just... sat on it for a while. If there's any confusing our unexplained changes, feel free to point them out and I'll try to give an explanation.
## Why It's Good For The Game

The goal of this PR is to give a bit more depth to security's armory implants. The intent is to present a choice in what implants are given (rather than just tracker and maybe chem if you're feeling spiteful), and to make them more useful as punitive/monitoring tools.

The tracker implant needed a nerf (and probably still does regardless of this PR's success). It's never used for tracking since the teleporter beacon is much more direct (+ gives a virtually free attack opportunity), and the tracking range was incredibly subpar. I'd rather not take toys away from security, but having the best option not be roundstart gear feels like a fair compromise.

Warden content. Wardens have more gear to budget for and use at their own (or the HOSes) discretion. The changes to the prisoner console allow them to coordinate with officers to get good value out of the implants they've chosen for an implantee. 

Gives antagonists an alternate way to get de-implanted, without external help, that can only be granted at the fault of security. Wardens who dish out implants must keep an eye on the people carrying them!
## Changelog
:cl: Rhials, MrMelbert
add: The Tracker implant has had its teleport beacon functionality migrated to the new (cargo accessible) Beacon implant.
add: Teleport Blocker security implant, that prevents the implantee from teleporting by any means. Purchasable from cargo.
add: Security implants may now be harmlessly self-destructed at the Prisoner Management Console.
balance: The Tracker implant tracking radius has increased from 20 to 35 tiles. The Prisoner Management Console will track and display the area the implantee is in as well.
balance: The exile implant now prevents implantees from operating shuttle controls.
code: Various code improvements and removal of unused vars in the Prisoner Management Console
code: The HUD slots for chem/tracking implants have been converted to display any implant with the IMPLANT_TYPE_SECURITY flag and an associated sprite.
spellcheck: Modifies various implant pad readouts, removing false information and rewriting some sections.
/:cl:

### CODE REVIEW

**Comment by Dawnseer** [comment]
How will this affect the tracker implant that miners can buy from the mining vendor? Will purchasing it give them both reds and yellows now?

**Comment by Profakos** [comment]
Could you please update the description of `/obj/item/ammo_box/c38/trac`? I had to expand it to clarify that unlike regular tracking implants, these do not allow teleportation, as people were confused, and this addition is no longer neccesary with this change. 👍

**Comment by Rhials** [comment]
> How will this affect the tracker implant that miners can buy from the mining vendor? Will purchasing it give them both reds and yellows now?

I totally meant to change this at some point but forgot to. Oops!

The miner tracker kit now gives two trackers and two beacons. If someone more willing to balance mining content wishes, they have the freedom to adjust what the kit gives.

> Could you please update the description of `/obj/item/ammo_box/c38/trac`? I had to expand it to clarify that unlike regular tracking implants, these do not allow teleportation, as people were confused, and this addition is no longer neccesary with this change. 👍

Donezo.

File: code/datums/components/tracked_implant.dm [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+/datum/component/tracked_implant/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_IMPLANT_IMPLANTED, PROC_REF(on_implant))
+	RegisterSignal(parent, COMSIG_IMPLANT_REMOVED, PROC_REF(on_remove))
+
+/datum/component/tracked_implant/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_IMPLANT_IMPLANTED, COMSIG_IMPLANT_REMOVED))
+
+/datum/component/tracked_implant/proc/on_implant(datum/source, mob/living/target, mob/user, silent = FALSE, force = FALSE)
+	global_list += parent
+	GLOB.tracked_generic_implants += parent
```
Comment (MrMelbert) [inline code/datums/components/tracked_implant.dm:26]:
```suggestion
/datum/component/tracked_implant/proc/on_implant(datum/source, mob/living/target, mob/user, silent = FALSE, force = FALSE)
	SIGNAL_HANDLER
	global_list += parent
	GLOB.tracked_generic_implants += parent
```

File: code/datums/components/tracked_implant.dm [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+/datum/component/tracked_implant/proc/on_remove(datum/source, mob/target, silent = FALSE, special = FALSE)
+	global_list -= parent
+	GLOB.tracked_generic_implants -= parent
```
Comment (MrMelbert) [inline code/datums/components/tracked_implant.dm:30]:
```suggestion
/datum/component/tracked_implant/proc/on_remove(datum/source, mob/target, silent = FALSE, special = FALSE)
	SIGNAL_HANDLER
	global_list -= parent
	GLOB.tracked_generic_implants -= parent
```

**Comment by GPeckman** [comment]
I do find the teleport blocker pretty cool. Could you keep an abductor agent trapped that way, if you managed to detain them and put it in fast enough?

**Comment by Dawnseer** [comment]
> I totally meant to change this at some point but forgot to. Oops!
> 
> The miner tracker kit now gives two trackers and two beacons, since I don't care to disrupt miners with my security changes. If someone more willing to balance mining content wishes, they have the freedom to adjust what the kit gives.

Cheers! Also is it possible to make a miner subset of the red? That doesn't decay after the 10 minutes? Because it can take a while for someone to notice that the miner is dead, At which point the point of the implant kinda dies if it just stops functioning

**Comment by DaxDupont** [comment]
> > I totally meant to change this at some point but forgot to. Oops!
> > The miner tracker kit now gives two trackers and two beacons, since I don't care to disrupt miners with my security changes. If someone more willing to balance mining content wishes, they have the freedom to adjust what the kit gives.
> 
> Cheers! Also is it possible to make a miner subset of the red? That doesn't decay after the 10 minutes? Because it can take a while for someone to notice that the miner is dead, At which point the point of the implant kinda dies if it just stops functioning

Yeah this is a concern of mine

**Comment by Rhials** [comment]
> Cheers! Also is it possible to make a miner subset of the red? That doesn't decay after the 10 minutes? Because it can take a while for someone to notice that the miner is dead, At which point the point of the implant kinda dies if it just stops functioning

It would be entirely possible to change, but out of the scope of the current PR. The 10 minute decay timer for tracking implants is not new behavior, and I don't want to muddy my security nerfbuff with an adjustment to miner gear that isn't strictly collateral.

**Comment by DrDiasyl** [comment]
The division of Tracking implant into both Tracking and Beacon is useless and unnecessary.
Security would just buy those roundstart and implant them both into the person, and they won't need to buy them against because you rarely implant that many people.

**Comment by Rhials** [comment]
> The division of Tracking implant into both Tracking and Beacon is useless and unnecessary. Security would just buy those roundstart and implant them both into the person, and they won't need to buy them again because you rarely implant that many people.

The teleporter beacon was the only aspect of the tracking implant that saw widespread use. Splitting them was necessary for the sake of balance, and personally I think it's a bit silly that the Tracker implant's tracking functionality got overshadowed so hard. My goal was to preserve as much of the currently present implant behavior/utility as possible while making these changes.

There is still the opportunity cost (one less implant slot) and aforementioned credit cost, but (by design) you can be back to how things were in only a short while. The intent was to have some consideration required for security to access the best option available, rather than it being roundstart gear.

**Comment by DrDiasyl** [comment]
> > The division of Tracking implant into both Tracking and Beacon is useless and unnecessary. Security would just buy those roundstart and implant them both into the person, and they won't need to buy them again because you rarely implant that many people.
> 
> The teleporter beacon was the only aspect of the tracking implant that saw widespread use. Splitting them was necessary for the sake of balance, and personally I think it's a bit silly that the Tracker implant's tracking functionality got overshadowed so hard. My goal was to preserve as much of the currently present implant behavior/utility as possible while making these changes.
> 
> There is still the opportunity cost (one less implant slot) and aforementioned credit cost, but (by design) you can be back to how things were in only a short while. The intent was to have some consideration required for security to access the best option available, rather than it being roundstart gear.

How is that balance?

**Comment by DATA-xPUNGED** [comment]
I don't really see how splitting off the tracker implant balances it in any way, just seems annoying

**Comment by Rhials** [comment]
> How is that balance?

The tracker and beacon functions being tied to the same implant is nonsensical since the beacon vastly overshadows the tracker in utility. The split allows for security to hold onto the tracker function (less utility, more effort, but roundstart gear) while having to put a tiny bit more work and consideration for the beacon function (more utility, less effort, not roundstart gear). The alternative is keeping them in the same implant and either nerfing the beacon part (wouldn't be received well at all) or buffing the tracker part (why would you do that when you have the beacon).

As per my last message, the intent was to have some consideration required for security to access the best option available, rather than it being roundstart gear. Making gear that is more powerful accessible later into the round is, as far as I'm aware, how every department in this game is balanced.

If this explanation isn't sufficient, try replying with more than four words so I can actually tell what I'm supposed to be arguing for. You're giving me nothing to work with here.

**Comment by JupiterJaeden** [comment]
Yea so, I already explained my position on this in manuelcord, but in essence, you haven't split the old tracker into "Tracker" and "Beacon". You have split it into "Tracker" and "Tracker Deluxe". There is really no reason for security to use the new tracker as soon as they can get their hands on beacons, because you can just use the tele console to check where somebody is. Even that aside, like you said yourself, the main functionality is the teleportation (and imo also the HUD identification for people concealing their identity). If your point was to just lock a straight upgrade behind not being roundstart gear, you have accomplished nothing but added another unnecessary thing to the checklist of roundstart prep for security.

If you really think the teleportation is unfair, just remove it and make the tracker pinpointer have unlimited range- but I think it's a little strange to claim trackers are too strong when you have left chemical implants 100% intact, which are literally a ranged 'you die' button. Frankly there is more broad counterplay to sec teleporting on top of you than sec smacking your chem implant (to which there exists precisely one form of counterplay, fast acting purge chemicals). 

Also, implants _are supposed_ to be a bit unfair- that was the point of security implanting you in the first place. It's supposed to be a serious disadvantage that you need to deal with if you intend to go loud, or risk it and suffer the potential consequences. In this sense, I *like* that you have added another route for antags to remove their implants other than 'hey doctor can you take this out' which we shouldn't be encouraging. But having added that, why do we still need to nerf implants. imo it would be extremely lame for manuel (and lets not pretend this isn't mostly a manuel thing as implants are used way more here) to develop a meta of antags letting themselves get caught early knowing they'll just be slapped with a useless implant (this even already happens sometimes and it's lame af). There should be some incentive for antags to be stealthy early on- if nothing else, let the detective have job content.

**Comment by ATH1909** [comment]
> Yea so, I already explained my position on this in manuelcord, but in essence, you haven't split the old tracker into "Tracker" and "Beacon". You have split it into "Tracker" and "Tracker Deluxe". There is really no reason for security to use the new tracker as soon as they can get their hands on beacons, because you can just use the tele console to check where somebody is.

In that case, we could just explicitly make the beacon implant a tracker deluxe implant and give it both the tracking and beacon functionalities, making it an upgrade to the roundstart version.

**Comment by optimumtact** [comment]
The teleport blocker probably needs some extra downside other than just cost/availability than, because otherwise people will use it to prevent themselves being teleporting.

e.g heretic sacs, abductors wizard spells.

Perhaps it should prevent the use of BCI's and similar, maybe the brain should have only 1 implant slot, so loyalty implants etc are lost.

**Comment by Rhials** [comment]
> Perhaps it should prevent the use of BCI's and similar, maybe the brain should have only 1 implant slot, so loyalty implants etc are lost.

Honestly I'm not against the idea of making security implants just collide with Mindshields, so they can't be used defensively. It wouldn't be a perfect solution but it would deal with the possible (and sometimes current) practice of security implanting themselves for defensive purposes.

Regardless, there probably are a few game-breaking instances where the noteleport implant could be a load of bullshit that do need addressing -- It will probably require something more fine tuned than total teleport immunity.

**Comment by Rhials** [comment]
no teleporting

Being teleported with a teleport blocker paralyzes you for 5 seconds.

This shouldn't interfere with heretic sacs (do_teleport force arg ignores teleport blocking). I would like to state that any cases where the implant interferes with the victim's ability to be a "target" should be put up as an issue.

The abductor teleport isn't actually a teleport, it uses forceMove, so the blocker implant has no effect. I did not realize this until yesterday.

So, currently, abductors cannot be defended against by the teleport blocker, but they also cannot be used to keep one imprisoned. I'm content with not doing anything about that, since all abductor teleports go through the teleport pad and I don't want to come up with a solution to split them in this PR.

**Comment by Rhials** [comment]
The Prisoner Management Console has been updated to use TGUI, courtesy of MrMelbert. Cool!

tgui

File: tgui/packages/tgui/interfaces/PrisonerManagement.tsx [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    'implantTab',
+    Object.keys(implantsByCategory)[0]
+  );
+
+  // Converts a category ("tracking implant") to a more readable format ("Tracking")
+  const formatCategory = (category: string) => {
+    return category
+      .split(' ')
+      .map((word) => word.charAt(0).toUpperCase() + word.slice(1))[0];
+  };
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PrisonerManagement.tsx:96]: try, when you can, to move pure functions outside of the component. specifically at 92, it recreates functions like `formatCategory` on each render. this isn't a tremendous issue here but good practice lest we die of a thousand cuts.

77 can be also made into an external function,`const implantsByCategory = sortImplants(implants) // or similar`. 

Generally a function should only stay within a component if it affects state directly

File: tgui/packages/tgui/interfaces/PrisonerManagement.tsx [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+              selected={implantTab === category}
+              onClick={() => setImplantTab(category)}>
+              {formatCategory(category)}
+            </Tabs.Tab>
+          ))}
+        </Tabs>
+      </Stack.Item>
+      <Stack.Item>
+        {implantsByCategory[implantTab] &&
+          implantsByCategory[implantTab].map((implant) => (
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PrisonerManagement.tsx:114]:
```suggestion
        {implantsByCategory[implantTab]?.map((implant) => (
```
OR better is to ensure implantsByCategory is always an array via `sortImplants`

File: tgui/packages/tgui/interfaces/PrisonerManagement.tsx [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+              <Button onClick={() => act('insert_id')}>No ID Inserted</Button>
+            </LabeledList.Item>
+          )}
+        </LabeledList>
+      </Stack.Item>
+      {!!id && (
+        <Stack.Item>
+          <NoticeBox>
+            Space Law recommends quotas of 100 points per minute they would
+            normally serve in the brig.
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/PrisonerManagement.tsx:161]:
```suggestion
            Space Law recommends quotas of 100 points per minute that they would
            normally serve in the brig.
```
this reads weirdly, sorry

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/PrisonerManagement.tsx (hunk 5 above)
Comment (MrMelbert) [reply tgui/packages/tgui/interfaces/PrisonerManagement.tsx:161]: This is copied directly from what the console currently says. 
Also, `per minute they would serve` sounds fine to me? I dunno.

File: code/game/objects/items/implants/implant.dm [hunk 6]
```diff
@@ -60,12 +64,18 @@
 	if(!force && !can_be_implanted_in(target))
 		return FALSE
 
-	for(var/X in target.implants)
-		var/obj/item/implant/other_implant = X
+	var/security_implants = 0 //Used to track how many implants with the "security" flag are in the user.
+	for(var/obj/item/implant/other_implant in target.implants)
```
Comment (Ghommie) [inline code/game/objects/items/implants/implant.dm:68]:
```suggestion
	for(var/obj/item/implant/other_implant as anything in target.implants)
```

File: code/game/objects/items/implants/implant.dm [hunk 7]
```diff
@@ -60,12 +64,18 @@
...
 		var/flags = SEND_SIGNAL(other_implant, COMSIG_IMPLANT_OTHER, args, src)
 		if(flags & COMPONENT_STOP_IMPLANTING)
 			UNSETEMPTY(target.implants)
 			return FALSE
+		if(!force)
+			if(other_implant.implant_flags & IMPLANT_TYPE_SECURITY)
+				security_implants++
+				if(security_implants >= SECURITY_IMPLANT_CAP) //We've found too many security implants in this mob, and will reject implantation by normal means
+					balloon_alert(user, "too many security implants!")
+					return FALSE
```
Comment (Ghommie) [inline code/game/objects/items/implants/implant.dm:78]:
```suggestion
		if(!force && (other_implant.implant_flags & IMPLANT_TYPE_SECURITY))
			security_implants++
			if(security_implants >= SECURITY_IMPLANT_CAP) //We've found too many security implants in this mob, and will reject implantation by normal means
				balloon_alert(user, "too many security implants!")
				return FALSE
```

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Pull request merged**